homeassistant/components/zeroconf/__init__.py

```diff
@@ -37,6 +46,33 @@
 
     await zeroconf.register_service(info)
 
+    async def new_service(service_type, name):
+        """"""
+        service_info = await zeroconf.get_service_info(service_type, name)
+        info = info_from_service(service_info)
+        print(info)
+
+        for domain in manifest.SERVICE_TYPES[service_type]:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    domain,
+                    context={'source': config_entries.SOURCE_DISCOVERY},
+                    data=info
+                )
+            )
+
+    def service_update(_, service_type, name, state_change):
+        """"""
+        from aiozeroconf import ServiceStateChange
+        print("Service %s of type %s state changed: %s" % (
+            name, service_type, state_change))
+
+        if state_change is ServiceStateChange.Added:
+            hass.async_create_task(new_service(service_type, name))
+
+    for service in manifest.SERVICE_TYPES:
+        ServiceBrowser(zeroconf, service, handlers=[service_update])
```

Comment:
Instead of doing it like this, could we also implement a catch all service handler and just check any incoming services? Because  `manifest.SERVICE_TYPES` will grow big.

Reply:
There are 21 services right now, and the service browser is pretty light weight. This is also how netdisco did it and how the library should be used. Maybe keep it for the moment and do this suggestion should it feel needed later on?

Reply:
The correct way would be to have a single `ServiceBrowser` listening for all notifications (no matter which service type it's coming from), and doing the checks in `service_update` to see whom that update should be delivered to. This has the following pros (w/o having looked at the implementation details of it):
* No need to use a file handle / socket for each service type, saves also some memory.
* (If wanted) we can add a debug messages telling that a non-claimed service type has been seen; can be helpful for reporting this information to developers in case they didn't know that their device can be autodetected.

Should removal and additions also be events instead of calling `new_service` directly, to allow integrations also to hook directly to this information (i.e., to be able to act when a device has been removed if wanted?).

Reply:
21 is now, but once we make discovery this easy, expect it to grow a lot bigger a lot faster. We need to build it to a way that scales.

Reply:
So there is a catch all function in aiozeroconf (ZeroconfServiceTypes). It uses one ServiceBrowser and listens to _services._dns-sd._udp.local. I haven't verified it but it seems proper. If you prefer I can change to using this one instead

Reply:
k